coderd/insights_test.go

```diff
@@ -2351,3 +2351,101 @@ func TestGenericInsights_RBAC(t *testing.T) {
+	t.Parallel()
+
+	t.Run("Success", func(t *testing.T) {
+		t.Parallel()
+
+		client, db := coderdtest.NewWithDatabase(t, nil)
+		coderdtest.CreateFirstUser(t, client)
+		ctx, cancel := context.WithTimeout(context.Background(), testutil.WaitShort)
+		t.Cleanup(cancel)
+
+		// Given: a deployment with many users
+
+		dbgen.User(t, db, database.User{
+			Email:     "user1@coder.com",
+			Username:  "user1",
+			CreatedAt: daysAgo(2),
+		})
+		dbgen.User(t, db, database.User{
+			Email:     "user2@coder.com",
+			Username:  "user2",
+			CreatedAt: daysAgo(2),
+		})
+		dbgen.User(t, db, database.User{
+			Email:     "user3@coder.com",
+			Username:  "user3",
+			CreatedAt: daysAgo(1),
+		})
+
+		// When: requesting the accumulated total of users by day
+		res, err := client.TotalUsersInsight(ctx, codersdk.TotalUsersInsightRequest{
+			StartTime: daysAgo(2),
+			EndTime:   today(),
+		})
+		require.NoError(t, err)
+
+		// Then: expect the correct number of dates and the growth of the accumulated total of users
+		require.Len(t, res, 3, "expect 3 dates of data")
+
+		// First day
+		require.Equal(t, res[0].Date, daysAgo(2).Format(time.DateOnly))
+		require.Equal(t, res[0].Total, uint64(2))
+
+		// Second day
+		require.Equal(t, res[1].Date, daysAgo(1).Format(time.DateOnly))
+		require.Equal(t, res[1].Total, uint64(3))
+
+		// Third day
+		require.Equal(t, res[2].Date, today().Format(time.DateOnly))
+		require.Equal(t, res[2].Total, uint64(4))
+	})
+
+	t.Run("DatesWithNoUserRegistration", func(t *testing.T) {
+		t.Parallel()
+
+		// Given: a deployment with no users getting created every day
+		client, db := coderdtest.NewWithDatabase(t, nil)
+		coderdtest.CreateFirstUser(t, client)
+		ctx, cancel := context.WithTimeout(context.Background(), testutil.WaitShort)
+		t.Cleanup(cancel)
+		dbgen.User(t, db, database.User{
+			Email:     "user1@coder.com",
+			Username:  "user1",
+			CreatedAt: daysAgo(2),
+		})
+
+		// When: requesting the accumulated total of users for dates with no users created
+		res, err := client.TotalUsersInsight(ctx, codersdk.TotalUsersInsightRequest{
+			StartTime: daysAgo(2).Truncate(24 * time.Hour),
+			EndTime:   today(),
+		})
+		require.NoError(t, err)
+
+		// Then: expect the correct number of dates. If there are no users created, the total should be 0.
+		require.Len(t, res, 3, "expect 3 dates of data")
+
+		// First day
+		require.Equal(t, res[0].Date, daysAgo(2).Format(time.DateOnly))
+		require.Equal(t, res[0].Total, uint64(1))
+
+		// Second day - no users created so should contain the same total as the previous day
+		require.Equal(t, res[1].Date, daysAgo(1).Format(time.DateOnly))
+		require.Equal(t, res[1].Total, uint64(1))
+
+		// Third day
+		require.Equal(t, res[2].Date, today().Format(time.DateOnly))
+		require.Equal(t, res[2].Total, uint64(2))
+	})
+}
+
+func daysAgo(days int) time.Time {
+	return time.Now().UTC().AddDate(0, 0, -days).Truncate(24 * time.Hour)
+}
+
+func today() time.Time {
```

Comment:
This function name is a bit confusing. I know it makes sense for the test but might be better to define it inside the parent test function instead of entire `coderd_test` package.